code/modules/mob/living/say.dm

```diff
@@ -106,45 +106,42 @@ var/list/department_radio_keys = list(
 				verb = "lisps"
 
 		if(T)
-			var/list/hearturfs = list()
-
-			for(var/I in hear(message_range, T))
-				if(istype(I, /mob/))
-					var/mob/M = I
-					listening += M
-					hearturfs += M.locs[1]
-					for(var/obj/O in M.contents)
-						if(O.flags_atom & USES_HEARING)
-							listening_obj |= O
-				else if(istype(I, /obj/))
-					var/obj/O = I
-					hearturfs += O.locs[1]
-					if(O.flags_atom & USES_HEARING)
-						listening_obj |= O
-
-			for(var/mob/M as anything in GLOB.player_list)
-				if((M.stat == DEAD || isobserver(M)) && M.client && M.client.prefs && (M.client.prefs.toggles_chat & CHAT_GHOSTEARS))
-					listening |= M
-					continue
-				if(M.loc && (M.locs[1] in hearturfs))
-					listening |= M
+			for(var/mob/M as anything in hearers(message_range, T))
+				listening += M
+				hearturfs += M.locs[1]
 
 		var/speech_bubble_test = say_test(message)
 		var/image/speech_bubble = image('icons/mob/effects/talk.dmi', src, "[bubble_type][speech_bubble_test]", FLY_LAYER)
 
 		var/not_dead_speaker = (stat != DEAD)
 		if(not_dead_speaker)
 			langchat_speech(message, listening, speaking)
+
+		// TODO: Improve by having a list of players with the pref on
+		for(var/mob/M as anything in GLOB.player_list)
+			if((M.stat == DEAD || isobserver(M)) && M.client && M.client.prefs && (M.client.prefs.toggles_chat & CHAT_GHOSTEARS))
+				listening |= M
+				continue
+			if(M.loc && (M.locs[1] in hearturfs))
+				listening |= M
+		
 		for(var/mob/M as anything in listening)
 			M.hear_say(message, verb, speaking, alt_name, italics, src, speech_sound, sound_vol)
 		overlays += speech_bubble
 
 		addtimer(CALLBACK(src, PROC_REF(remove_speech_bubble), speech_bubble), 3 SECONDS)
 
-		for(var/obj/O as anything in listening_obj)
-			if(O) //It's possible that it could be deleted in the meantime.
+		for(var/obj/O as anything in GLOB.hearing_objects)
+			if(O && ((O in hearturfs) || (O.locs[1] in hearturfs))) //It's possible that it could be deleted in the meantime.
 				O.hear_talk(src, message, verb, speaking, italics)
 
+
+		for(var/obj/O as anything in GLOB.special_hearing_objects)
+			if(!O)
+				LAZYREMOVE(GLOB.special_hearing_objects, O)
+				continue
+			if(isInSight(O, T) || (O in hearturfs) || (isInSight(O.locs[1], T)))
```

Comment:
Byond uses short-circuiting for boolean (see https://www.byond.com/docs/ref/#/operator/|| ) so it would likely be better to first test if O is in our list before performing a line of sight check.